pkg/telemetry/telemetry.go

```diff
@@ -0,0 +1,118 @@
+package telemetry
+
+import (
+	"os"
+	"time"
+
+	metrics "github.com/armon/go-metrics"
+	"github.com/armon/go-metrics/datadog"
+)
+
+const (
+	noHostName  = ""
+	serviceName = "spiffe-aws-assume-role"
+
+	calls   = "Calls"
+	failure = "Failure"
+	latency = "Latency"
+	success = "Success"
+)
+
+type Telemetry struct {
+	Metrics *metrics.Metrics
+	labels  []metrics.Label
+}
+
+func (t *Telemetry) AddLabel(name string, value string) {
+	t.labels = append(t.labels, metrics.Label{Name: name, Value: value})
+}
+
+func NullTelemetry() (*Telemetry, error) {
+	return NewTelemetryForSinkAndHostname(&metrics.BlackholeSink{}, noHostName)
+}
+
+func MustNullTelemetry() *Telemetry {
+	telemetry, err := NullTelemetry()
+	if err != nil {
+		panic(err)
+	}
+	return telemetry
+}
+
+func NewTelemetry(address string) (*Telemetry, error) {
+	sink, err := newSink(address)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewTelemetryForSink(sink)
+}
+
+func NewTelemetryForSink(sink metrics.MetricSink) (*Telemetry, error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		hostname = ""
+	}
+
+	return NewTelemetryForSinkAndHostname(sink, hostname)
+}
+
+func NewTelemetryForSinkAndHostname(sink metrics.MetricSink, hostname string) (*Telemetry, error) {
+	_metrics, err := metrics.New(metrics.DefaultConfig(serviceName), sink)
+	if err != nil {
+		return nil, err
+	}
+	_metrics.EnableHostname = false
+
+	var labels []metrics.Label
+	if len(hostname) > 0 {
+		labels = append(labels, *newLabel("hostname", hostname))
+	}
+	telemetry := Telemetry{
+		Metrics: _metrics,
+		labels:  labels,
+	}
+
+	return &telemetry, nil
+}
+
+func newSink(address string) (metrics.MetricSink, error) {
+	if address == "" {
+		return &metrics.BlackholeSink{}, nil
+	} else {
+		return datadog.NewDogStatsdSink(address, noHostName)
+	}
+}
+
+func (t *Telemetry) Instrument(key []string, err *error) func() {
+	start := time.Now()
+
+	return func() {
+		latencyInMilliseconds := time.Since(start).Milliseconds()
+
+		t.Metrics.IncrCounterWithLabels(copyAndAppend(key, calls), 1, t.labels)
+		t.Metrics.SetGaugeWithLabels(copyAndAppend(key, latency), float32(latencyInMilliseconds), t.labels)
+
+		if *err == nil {
+			t.Metrics.IncrCounterWithLabels(copyAndAppend(key, success), 1, t.labels)
+			t.Metrics.IncrCounterWithLabels(copyAndAppend(key, failure), 0, t.labels)
+		} else {
+			t.Metrics.IncrCounterWithLabels(copyAndAppend(key, success), 0, t.labels)
+			t.Metrics.IncrCounterWithLabels(copyAndAppend(key, failure), 1, t.labels)
+		}
```

Comment:
Hopefully this covers all the bases.

Reply:
I renamed key to metricNamePrefix, hopefully that's more clear.

Reply:
So the "name" of a metric is hierarchichal and appears to be defined as an array of strings which is referred to as a "key" in the metrics library code (it's not the best name). In any case, what we have is a sort of prefix in the form of a three-part key ["spiffe-aws-assume-role", className, methodName] and what we want to do is append Success, Failure, Calls, Latency, etc. to the end of it. I'm not a wizard at Go arrays but per my reading we have to create a copy of the array before any append if we want it to be independent of other appends, and the syntax for that is kind of annoying. So, what I did is create that helper method to take care of it.

Reply:
Got it, thanks for the explanation. Looks like `copyAndAppend` does work (since it does what it's intended!), and I was about to comment "why not just use append instead?" but after some googling, it looks like `append` can have some weird behavior: https://medium.com/@Jarema./golang-slice-append-gotcha-e9020ff37374. So copying and appending does look like the right solution here!